exonum-java-binding/core/rust/src/testkit/mod.rs

```diff
@@ -213,3 +216,140 @@ fn create_java_keypair<'a>(
         &[secret_key_byte_array.into(), public_key_byte_array.into()],
     )
 }
+
+// Converts Java array of `TestKitServiceInstances` to vector of `InstanceConfig`.
+//
+// `TestKitServiceInstances` representation:
+//      String artifactId;
+//      String artifactFilename;
+//      ServiceSpec[] serviceSpecs;
+fn instance_configs_from_java_array(
+    env: &JNIEnv,
+    service_artifact_specs: jobjectArray,
+) -> JniResult<Vec<InstanceConfig>> {
+    let mut instance_configs = vec![];
+    let num_artifacts = env.get_array_length(service_artifact_specs)?;
+    for i in 0..num_artifacts {
+        let artifact_spec_obj =
+            env.auto_local(env.get_object_array_element(service_artifact_specs, i)?);
+
+        let artifact_id = to_artifact_id_str(get_field_as_string(
+            env,
+            artifact_spec_obj.as_obj(),
+            "artifactId",
+        )?);
+        let artifact_filename =
+            get_field_as_string(env, artifact_spec_obj.as_obj(), "artifactFilename")?;
+        let service_specs_obj: jobjectArray = env
+            .get_field(
+                artifact_spec_obj.as_obj(),
+                "serviceSpecs",
+                SERVICE_SPECS_FIELD_TYPE,
+            )?
+            .l()?
+            .into_inner();
+        let configs = parse_service_specs(env, service_specs_obj, artifact_id, artifact_filename)?;
+        instance_configs.extend(configs);
+    }
+    Ok(instance_configs)
+}
+
+// Converts Java array of `ServiceSpec` instances into vector of `InstanceConfig` for specific artifact.
+fn parse_service_specs(
+    env: &JNIEnv,
+    specs_array: jobjectArray,
+    artifact_id: String,
+    artifact_filename: String,
+) -> JniResult<Vec<InstanceConfig>> {
+    let num_specs = env.get_array_length(specs_array)?;
+
+    let mut instance_configs = vec![];
+    for i in 0..num_specs {
+        let service_spec = env.auto_local(env.get_object_array_element(specs_array, i)?);
+        let (spec, config) = parse_instance_spec(&env, service_spec.as_obj(), &artifact_id)?;
+        let cfg = InstanceConfig::new(spec, Some(artifact_filename.to_bytes()), config);
```

Comment:
@MakarovS The Java parameters are incorrect: the Java code must pass deploy arguments serialized
as a protobuf message, not the filename: exonum-java-binding/core/src/main/proto/service_runtime.proto:46

This code produces incorrect deploy arguments.

Reply:
Thanks, will fix that on Java side.